crypto/crmf/crmf_lib.c

```diff
@@ -612,102 +646,299 @@ int OSSL_CRMF_CERTTEMPLATE_fill(OSSL_CRMF_CERTTEMPLATE *tmpl,
-                                       OSSL_LIB_CTX *libctx, const char *propq,
-                                       EVP_PKEY *pkey)
+#ifndef OPENSSL_NO_CMS
+DECLARE_ASN1_ITEM(CMS_SignedData) /* copied from cms_local.h */
+
+/* check for KGA authorization implied by CA flag or by explicit EKU cmKGA */
+static int check_cmKGA(ossl_unused const X509_PURPOSE *purpose, const X509 *x, int ca)
+{
+    STACK_OF(ASN1_OBJECT) *ekus;
+    int i, ret = 1;
+
+    if (ca)
+        return ret;
+    ekus = X509_get_ext_d2i(x, NID_ext_key_usage, NULL, NULL);
+    for (i = 0; i < sk_ASN1_OBJECT_num(ekus); i++) {
+        if (OBJ_obj2nid(sk_ASN1_OBJECT_value(ekus, i)) == NID_cmKGA)
+            goto end;
+    }
+    ret = 0;
+
+ end:
+    sk_ASN1_OBJECT_pop_free(ekus, ASN1_OBJECT_free);
+    return ret;
+}
+#endif /* OPENSSL_NO_CMS */
+
+EVP_PKEY *OSSL_CRMF_ENCRYPTEDKEY_get1_pkey(const OSSL_CRMF_ENCRYPTEDKEY *encryptedKey,
+                                           X509_STORE *ts, STACK_OF(X509) *extra, EVP_PKEY *pkey,
+                                           X509 *cert, ASN1_OCTET_STRING *secret,
+                                           OSSL_LIB_CTX *libctx, const char *propq)
+{
+#ifndef OPENSSL_NO_CMS
+    BIO *bio = NULL;
+    CMS_SignedData *sd = NULL;
+    BIO *pkey_bio = NULL;
+    int purpose_id, bak_purpose_id;
+    X509_VERIFY_PARAM *vpm;
+#endif
+    EVP_PKEY *ret = NULL;
+
+    if (encryptedKey == NULL) {
+        ERR_raise(ERR_LIB_CRMF, CRMF_R_NULL_ARGUMENT);
+        return NULL;
+    }
+    if (encryptedKey->type != OSSL_CRMF_ENCRYPTEDKEY_ENVELOPEDDATA) {
+        unsigned char *p;
+        const unsigned char *p_copy;
+        int len;
+
+        p = OSSL_CRMF_ENCRYPTEDVALUE_decrypt(encryptedKey->value.encryptedValue,
+                                             libctx, propq, pkey, &len);
+        if ((p_copy = p) != NULL)
+            ret = d2i_AutoPrivateKey_ex(NULL, &p_copy, len, libctx, propq);
+        OPENSSL_free(p);
+        return ret;
+    }
+
+#ifndef OPENSSL_NO_CMS
+    if (ts == NULL) {
+        ERR_raise(ERR_LIB_CRMF, CRMF_R_NULL_ARGUMENT);
+        return NULL;
+    }
+    if ((bio = CMS_EnvelopedData_decrypt(encryptedKey->value.envelopedData,
+                                         NULL, pkey, cert, secret, 0,
+                                         libctx, propq)) == NULL) {
+        ERR_raise(ERR_LIB_CRMF, CRMF_R_ERROR_DECRYPTING_ENCRYPTEDKEY);
+        goto end;
+    }
+    sd = ASN1_item_d2i_bio(ASN1_ITEM_rptr(CMS_SignedData), bio, NULL);
+    if (sd == NULL)
+        goto end;
+
+    if ((purpose_id = X509_PURPOSE_get_by_sname(SN_cmKGA)) < 0) {
+        purpose_id = X509_PURPOSE_get_unused_id(libctx);
+        if (!X509_PURPOSE_add(purpose_id, X509_TRUST_COMPAT, 0, check_cmKGA,
+                              LN_cmKGA, SN_cmKGA, NULL)) {
+            ERR_raise(ERR_LIB_CRMF, CRMF_R_ERROR_SETTING_PURPOSE);
+            goto end;
+        }
```

Comment:
```suggestion
                              LN_cmKGA, SN_cmKGA, NULL))
            goto end;
```
The `CRMF_R_ERROR_SETTING_PURPOSE` is not really adequate here
and it is not needed because `X509_PURPOSE_add()` add and needed error queue entry itself.